### ISSUE TITLE
float numbers error in garmin_connect.add_body_composition service

### ISSUE THREAD

**Issue by pejotigrek** [cyberjunky/home-assistant-garmin_connect#127]
first of all - it is GREAT to see the `garmin_connect.add_body_composition` service! I am starting to use it more and more [manually right now, but automation is on the todo list :)] with data from my Xiaomi MiScale.

while playing with the service I've noticed it has some problems with numbers that can have float values, but they don't - for example:
adding `percent_hydration: 54.0` gives this error:

```
expected float for dictionary value @ data['percent_hydration']. Got None
```

but when I enter `percent_hydration: 54.1` - all is fine. just like when the number has no decimals [or has 0 after the decimal point] service fails to work.

the very same error was with all parameters. if those values were integer [*.0] it failed. when I've changed *.0 to any other digit - service added all data to Garmin.

**Comment by Cul8rch** [comment]
Oh no....yesterday I spent hours on a node red flow to pull data from Fitbit and feed them to Garmin, failing on values with *.0 . The above described error is exactly the same I got and therefore seems to still be open and unsolved. Is there at least a workaround on that? As it is now, the add_body_composition is useless.

**Comment by pejotigrek** [comment]
> (...) Is there at least a workaround on that? As it is now, the add_body_composition is useless.

I'm not using NR for my automations, but did some workaround in yaml/native automation for HA I guess you can use the same idea in NR: before I send anything to Garmin, I'm testing if measurements are floating or integers. if the number has some decimals then it's sent as it is; if the number is integer, I just add 0.1 to it and send it like a float. I know that this "cheats" the scores a little bit [example: instead of 11% body fat, Garmin gets 11.1%, instead of 67% hydration Garmin gets 67.1% etc.] but this is a sacrifice I am ready to take ;)

**Comment by Cul8rch** [comment]
Thanks @pejotigrek for the fast reply. I had that also in mind as a dirty workaround, and in the case of Weight or Bodyfat it's in general positive if I go to 11.9 instead 12.0 ;-)
In the meantime, I can say, that it is not a problem of the node red or automation script as the error is reproducible with manual execution in the actions tab with for example:
action: garmin_connect.add_body_composition
data:
  entity_id: sensor.weight
  timestamp: "2025-01-04T09:06:45.235Z"
  weight: 82.0 

The sourcecode expects float as far as I can say:
BODY_COMPOSITION_SERVICE_SCHEMA = vol.Schema(
    {
        vol.Required(ATTR_ENTITY_ID): str,
        vol.Optional("timestamp"): str,
        vol.Required("weight"): float,
        vol.Optional("percent_fat"): float,
        vol.Optional("percent_hydration"): float,
        
So could it be that this problem is somewhere in the service / action handling of Homeassistant?

**Comment by pejotigrek** [comment]
yes, the same behavior was in my case - no matter which method is used to update, the error is showing up.
I guess it is something with value conversion, so you're right probably - but unfortunately this is where my knowledge about programming is starting to fade ;)

**Comment by Cul8rch** [comment]
ok figured out a solution based on your cheat, which does not affect the effective value. Just add 0.000001 to each value. It would have only an impact if your original value was *.999999 and that will occur veeeeeery seldom. But for 80.0 it should do the trick.

**Comment by pejotigrek** [comment]
@Cul8rch well.. that's perfect!
haha, just because I don't care with measures less than 1/10th I just stopped by adding 0.1 ... and your solution is much better by adding a few zeroes :) thanks, fixing my scripts as we speak ;)

**Comment by cyberjunky** [comment]
All, I have found a solution so both int and floats are accepted...
```vol.Required(ATTR_ENTITY_ID): str,
            vol.Optional("timestamp"): str,
            vol.Required("weight"): vol.Coerce(float),
            vol.Optional("percent_fat"): vol.Coerce(float)

```
Also sorted out the deprecated service schema error, but had to move a lot of things around,  still need to test/look at the gear service call, when that's working again I will release update. Perhaps it even fixes the multiple account add body composition issue, but not sure yet.

**Comment by cyberjunky** [comment]
Implemented solution which accepts both ints and floats in latest versions, so closing, no need to add the 0.00001 values anymore
